msdm/algorithms/entregpolicyiteration.py

```diff
@@ -0,0 +1,153 @@
+    transition_matrix: torch.Tensor,
+    reward_matrix : torch.Tensor,
+    discount_rate : float,
+    entropy_weight : Union[torch.Tensor, float],
+    n_planning_iters : int,
+    policy_prior : torch.Tensor=None,
+    initial_policy : torch.Tensor=None,
+    check_convergence : bool=True,
+    q_value_range : float=None
+):
+    """
+    An implementation of entropy regularized policy iteration
+    (e.g., Geist, Scherrer, & Pietquin (2019) ICML). This takes in
+    pytorch tensors and returns tensors that result from the computation.
+
+    Parameters:
+        transition_matrix: An SxAxS tensor representing MDP transition probabilities
+        reward_matrix: An (1|S)x(1|A)x(1|S) tensor representing MDP rewards
+        discount_rate: The discount rate
+        entropy_weight: Weight(s) on the entropy regularization. When a S-length vector
+            is passed in, each entry corresponds to the weight on that state.
+        n_planning_iters: Iterations of policy iteration (improvement and evaluation)
+        policy_prior: An (1|S)xA tensor representing policy prior probabilities
+        initial_policy: An SxA tensor representing the initial policy for policy iteration
+        check_convergence: Whether to check convergence
+        stop_on_convergence: Whether to stop at convergence
+        q_value_range: To prevent underflow when softmaxing, this lets you set
+            a floor on how much lower scaled values can be than the maximum.
+
+    Returns:
+        A SimpleNamespace with the results of the algorithm.
+    """
+    tf = transition_matrix
+    rf = reward_matrix
+    if policy_prior is None:
+        policy_prior = torch.softmax(torch.ones((1, tf.shape[1]), dtype=tf.dtype), 1)
+    if initial_policy is None:
+        initial_policy = torch.softmax(torch.ones(tf.shape[:-1], dtype=tf.dtype), 1)
+    assert isinstance(transition_matrix, torch.Tensor)
+    assert initial_policy.shape == tf.shape[:-1]
+    pi0 = policy_prior
+    assert tf.shape[0] == tf.shape[2]
+    assert rf.shape[0] in (1, tf.shape[0])
+    assert rf.shape[1] in (1, tf.shape[1])
+    assert rf.shape[2] in (1, tf.shape[2])
+    assert pi0.shape[0] in (1, tf.shape[0])
+    assert pi0.shape[1] == tf.shape[1]
+    if isinstance(entropy_weight, (float, int)):
+        entropy_weight = torch.tensor([entropy_weight])
+    assert entropy_weight.shape[0] in (1, tf.shape[0])
+
+    eye = torch.eye(tf.shape[0])
+    pi = initial_policy
+    converged = False
+    for i in range(n_planning_iters):
+        s_ent = torch.nansum(torch.log(pi/pi0)*pi, dim=1)
+        s_rf = torch.einsum("san,san,sa->s",rf,tf,pi)
+        s_rf_ent = (s_rf - entropy_weight*s_ent)
+        mp = (pi[:,:,None]*tf[:, :, :]).sum(dim=1)
+        v = torch.linalg.solve(eye - discount_rate*mp, s_rf_ent)
+        q = (tf[:,:,:]*(rf + v[None,None,:])).sum(dim=-1)
+        q_scale = q_action = (1/entropy_weight[:,None])*q
+        if q_value_range is not None:
```

Comment:
Wonder if this & what follows should be pulled out into a utility for logsoftmax?